File: Depth_First_Search/1315.Sum_of_Nodes_with_Even-Valued_Grandparent/Python/1315.Sum_of_Nodes_with_Even-Valued_Grandparent.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def sumEvenGrandparent(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#

        ############
        #Initialize
        ##### Result summary value #####
        res_sum_val = [0]

        ####################################
        #Recursion-based DFS loop traversal
        self.dfsTraversal(root, res_sum_val) #Recursion function call (i.e. whole root-node)

        return res_sum_val[0]


    def dfsTraversal(self, root, res_sum_val):
        """
        :type root: Optional[TreeNode]
        :type res_sum_val: List[int]
        :rtype: None, void
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#

        ####################
        #Whole process/flow
        ##### Check if the current child-nodes existed or not #####
        if (root.left):

            ##### Check if the current grand child-nodes existed or not #####
            if ((root.left).left):

                ##### Check if the current grand parent-node is even or not #####
                if (not (root.val % 2)):
                    res_sum_val[0] += (((root.left).left).val) #Keep updating/accumulating

            if ((root.left).right):

                ##### Check if the current grand parent-node is even or not #####
                if (not (root.val % 2)):
                    res_sum_val[0] += (((root.left).right).val) #Keep updating/accumulating

            self.dfsTraversal(root.left, res_sum_val) #Recursion function call (i.e. root left-half node)

        if (root.right):

            ##### Check if the current grand child-nodes existed or not #####
            if ((root.right).left):

                ##### Check if the current grand parent-node is even or not #####
                if (not (root.val % 2)):
                    res_sum_val[0] += (((root.right).left).val) #Keep updating/accumulating

            if ((root.right).right):

                ##### Check if the current grand parent-node is even or not #####
                if (not (root.val % 2)):
                    res_sum_val[0] += (((root.right).right).val) #Keep updating/accumulating

            self.dfsTraversal(root.right, res_sum_val) #Recursion function call (i.e. root right-half node)

File: Depth_First_Search/1315.Sum_of_Nodes_with_Even-Valued_Grandparent/C++/1315.Sum_of_Nodes_with_Even-Valued_Grandparent.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumEvenGrandparent(TreeNode * root) 
    {
        //=================================//
        // Input type:                     //
        //   - TreeNode * (root tree-node) //
        // Return type:                    //
        //   - int (result summary value)  //
        //=================================//

        /*Initialize*/
        ///// Result summary value /////
        int res_sum_val = 0;

        /*Recursion-based DFS loop traversal*/
        dfsTraversal(root, res_sum_val); //Recursion function call (i.e. whole root-node)

        return res_sum_val;
    }


    void dfsTraversal(TreeNode * root, int & res_sum_val)
    {
        //=====================================//
        // Input type:                         //
        //   - TreeNode * (root tree-node)     //
        // Return type:                        //
        //   - int ref. (result summary value) //
        //=====================================//

        /*Whole process, flow*/
        ///// Check if the child-nodes existed or not /////
        if ((root->left) != nullptr)
        {
            ///// Check if the child-nodes existed or not /////
            if (((root->left)->left) != nullptr)
            {
                ///// Check if the current grand-parent node is even or not /////
                if (! ((root->val) % 2)) { res_sum_val += (((root->left)->left)->val); } //Keep updating, accumulating
            }

            if (((root->left)->right) != nullptr)
            {
                ///// Check if the current grand-parent node is even or not /////
                if (! ((root->val) % 2)) { res_sum_val += (((root->left)->right)->val); } //Keep updating, accumulating
            }

            dfsTraversal((root->left), res_sum_val); //Recursion function call (i.e. root left-half node)
        }

        if ((root->right) != nullptr)
        {
            ///// Check if the child-nodes existed or not /////
            if (((root->right)->left) != nullptr)
            {
                ///// Check if the current grand-parent node is even or not /////
                if (! ((root->val) % 2)) { res_sum_val += (((root->right)->left)->val); } //Keep updating, accumulating
            }

            if (((root->right)->right) != nullptr)
            {
                ///// Check if the current grand-parent node is even or not /////
                if (! ((root->val) % 2)) { res_sum_val += (((root->right)->right)->val); } //Keep updating, accumulating
            }

            dfsTraversal((root->right), res_sum_val); //Recursion function call (i.e. root right-half node)
        }
    }
};
